lib/Controller/DirectUploadController.php

```diff
@@ -167,12 +171,26 @@ public function directUpload(string $token):DataResponse {
 			if (empty($nodes)) {
 				throw new NotFoundException('folder not found or not enough permissions');
 			}
+			/**
+			 * @var Folder $folderNode
+			 */
 			$folderNode = array_shift($nodes);
+			$freeSpace = $folderNode->getFreeSpace();
+
+			// this is also true if we try to overwrite
+			// to overwrite a file we need enough free quota for the new data
+			// otherwise `putContent()` fails
+			if ($directUploadFile['size'] > $freeSpace) {
+				throw new NotEnoughSpaceException('insufficient quota');
+			}
 			if (
 				$folderNode->isCreatable()
 			) {
 				// @phpstan-ignore-next-line
 				if ($folderNode->nodeExists($fileName) && $overwrite) {
+					/**
+					 * @var File $file
+					 */
 					$file = $folderNode->get($fileName); // @phpstan-ignore-line
 					if ($file->getType() === FileInfo::TYPE_FOLDER) {
```

Comment:
Maybe these can be removed now?
```suggestion
					$file = $folderNode->get($fileName);
					if ($file->getType() === FileInfo::TYPE_FOLDER) {
```

Reply:
after #328 I will make a separate PR that will hopefully get rid of all the `@phpstan-ignore-line` bits